text/0000-zero-sized-references.md

```diff
@@ -0,0 +1,142 @@
+a warning/error be issued ("Warning: taking the address of a Zero-Sized Type is meaningless") and the pointer will recieve an address
+with the same algorithm that assigned an address to the reference in our current implementation.
+
+The purpose is to not break current code that might do this. We probably don't want to assign Null since for pointers it has
+a meaning that the value doesn't exist, which is different than "exist but no data".
+
+Converting in the other direction, the value of the pointer will be silently dropped - that value never had a meaning in the first place.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+For most rust users, this change will be invisible. Thier code will just become a tiny bit smaller.
+
+Users of unsafe rust might encounter this case. Therefore there should probably be a note in the nomicon that references might be optimized away for ZST.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### Breaking code
+
+Any place that assumed a reference holds a pointer might introduce bugs.
+
+FFI might behave differently than now, breaking code (see Unresolved questions).
+
+# Alternatives
+[alternatives]: #alternatives
+
+The system that is now works well, and does not have to be changed.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+### FFI + Escape mechanism
+
+Do we need a way to remove the optimization? For example, if we have
+```rust
+struct ExternalStruct;
+
+let x: &mut ExternalStruct = ffi_function();
+other_ffi_function(x);
+```
+We might want to represent a pointer we got from FFI and checked it's not null as a reference to an object.
+Since that object is not accessable directly we represent it as an empty struct.
+However then the references to it are optimized away and we lose the pointer and then can't call the second function.
+
+Note: This break will not be silent. The safe wrapper for other_ffi_function will convert the reference back to pointer before
+using it, raising the warning/error of converting references to pointers.
+
+For that case, we might want to mark `ExternalStruct` as Non-Zero-Sized. Possibly
+```rust
+impl !Sized for ExternalStruct;
+```
+and maybe it should require `unsafe`.
+
+### Errors of conversion
+
+Do we want to give an error or a warning for converting references to pointers?
+
+The code above is an example where we break working code, so an error might be needed to show the significance.
+
+However, some conversions might be meaningless and wouldn't affect the execution, so the programmer might allow the conversion.
+
+### Mitigating breakage
+
+Safe rust code should be affected positively by this change. However unsafe code might break.
+
+FFI is especially vulnurable to this change, as shown above.
+Are there better ways to deal with these errors without user involvement?
+
+### Specific examples - Pro
+
+The RFC isn't well-justified until it has at least one detailed use case where it helps.
```

Comment:
My opinion is that it's enough justification that this will solve all the issues around addresses and zsts once and for all.